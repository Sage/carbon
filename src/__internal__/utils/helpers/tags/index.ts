export { default } from "./tags";
export type { TagProps } from "./tags";
