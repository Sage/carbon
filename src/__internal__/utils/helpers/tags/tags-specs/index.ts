export { elementsTagTest, rootTagTest } from "./tags-specs";
