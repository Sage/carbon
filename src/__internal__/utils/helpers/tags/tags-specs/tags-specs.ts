import { ReactWrapper, ShallowWrapper } from "enzyme";

const elementsTagTest = (wrapper: ReactWrapper, elements: string[]) => {
  elements.forEach((element) => {
    it(`include 'data-element="${element}"'`, () => {
      expect(wrapper.find({ "data-element": element }).length).toEqual(1);
    });
  });
};

const rootTagTest = (
  rootNode: ReactWrapper | ShallowWrapper,
  comp: string,
  elem?: string,
  role?: string
) => {
  expect(rootNode.prop("data-component")).toEqual(comp);
  expect(rootNode.prop("data-element")).toEqual(elem);
  expect(rootNode.prop("data-role")).toEqual(role);
};

// eslint-disable-next-line jest/no-export
export { elementsTagTest, rootTagTest };
