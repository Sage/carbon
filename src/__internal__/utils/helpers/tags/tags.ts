interface DataProps {
  /** Identifier used for testing purposes, applied to the root element of the component. */
  "data-element"?: string;
  /** Identifier used for testing purposes, applied to the root element of the component. */
  "data-role"?: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [restKeys: string]: any;
}

export interface TagProps extends DataProps {
  /** Identifier used for testing purposes, applied to the root element of the component. */
  "data-component"?: string;
}

/**
 * Builds props object containing top level data tags
 */
function tagComponent(
  componentName: string | undefined,
  props: DataProps
): TagProps {
  const tagProps: TagProps = {
    "data-component": componentName,
  };

  if (props["data-element"]) {
    tagProps["data-element"] = props["data-element"] as string;
  }
  if (props["data-role"]) {
    tagProps["data-role"] = props["data-role"] as string;
  }

  return tagProps;
}

export default tagComponent;
