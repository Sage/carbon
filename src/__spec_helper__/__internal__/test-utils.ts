/* eslint-disable jest/no-conditional-expect */
/* eslint-disable jest/no-identical-title */
/* eslint-disable jest/no-export */
import { render } from "@testing-library/react";
import { sprintf } from "sprintf-js";
import {
  LayoutProps,
  FlexboxProps,
  BackgroundProps,
  PositionProps,
  GridProps,
} from "styled-system";

import { space } from "../../style/themes/base/base-theme.config";
import { mockMatchMedia } from "../mock-match-media";

const isUpper = (char: string) => char.toUpperCase() === char;
const humpToDash = (acc: string, char: string) =>
  `${acc}${isUpper(char) ? `-${char.toLowerCase()}` : char}`;

const toCSSCase = (str: string) => {
  return str.split("").reduce(humpToDash, "");
};

const assertStyleMatch = (
  styleSpec: { [key: string]: string | number | undefined },
  component: HTMLElement,
  opts?: jest.Options,
) => {
  Object.entries(styleSpec).forEach(([attr, value]) => {
    expect(component).toHaveStyleRule(toCSSCase(attr), value, opts);
  });
};

const makeArrayKeys = (n: number) => [...Array(n).keys()];

const marginProps = [
  ["m", "margin"],
  ["ml", "marginLeft"],
  ["mr", "marginRight"],
  ["mt", "marginTop"],
  ["mb", "marginBottom"],
  ["mx", "marginLeft"],
  ["mx", "marginRight"],
  ["my", "marginTop"],
  ["my", "marginBottom"],
] as const;

type MarginProps = {
  [K in (typeof marginProps)[number][0]]?: string | number;
};

const paddingProps = [
  ["p", "padding"],
  ["pl", "paddingLeft"],
  ["pr", "paddingRight"],
  ["pt", "paddingTop"],
  ["pb", "paddingBottom"],
  ["px", "paddingLeft"],
  ["px", "paddingRight"],
  ["py", "paddingTop"],
  ["py", "paddingBottom"],
] as const;

type PaddingProps = {
  [K in (typeof paddingProps)[number][0]]?: string | number;
};

const colorProps = [
  ["color", "color", "#CCCCCC"],
  ["bg", "background-color", "#FFFFFF"],
  ["opacity", "opacity", "0.5"],
] as const;

const widthProps = ["width", "width", "200px"] as const;

const heightProps = ["height", "height", "200px"] as const;

const layoutProps = [
  widthProps,
  heightProps,
  ["minWidth", "min-width", "120px"],
  ["maxWidth", "max-width", "120px"],
  ["minHeight", "min-height", "120px"],
  ["maxHeight", "max-height", "120px"],
  ["size", "width", "120px"],
  ["size", "height", "120px"],
  ["display", "display", "inline-block"],
  ["verticalAlign", "vertical-align", "baseline"],
  ["overflow", "overflow", "hidden"],
  ["overflowX", "overflow-x", "hidden"],
  ["overflowY", "overflow-y", "hidden"],
] as const;

const flexBoxProps = [
  ["alignItems", "alignItems", "center"],
  ["alignContent", "alignContent", "center"],
  ["justifyItems", "justifyItems", "center"],
  ["justifyContent", "justifyContent", "center"],
  ["flexWrap", "flexWrap", "wrap"],
  ["flexDirection", "flexDirection", "row-reverse"],
  ["flex", "flex", "1"],
  ["flexGrow", "flexGrow", "1"],
  ["flexShrink", "flexShrink", "1"],
  ["flexBasis", "flexBasis", "100px"],
  ["justifySelf", "justifySelf", "center"],
  ["alignSelf", "alignSelf", "center"],
  ["order", "order", "1"],
] as const;

const gridProps = [
  ["gridColumn", "gridColumn", "1 / span 2"],
  ["gridRow", "gridRow", "1 / span 2"],
  ["gridArea", "gridArea", "myArea"],
  ["gridAutoFlow", "gridAutoFlow", "column"],
  ["gridAutoRows", "gridAutoRows", "150px"],
  ["gridAutoColumns", "gridAutoColumns", "50px"],
  ["gridTemplateRows", "gridTemplateRows", "100px 300px"],
  ["gridTemplateColumns", "gridTemplateColumns", "auto auto auto auto"],
  ["gridTemplateAreas", "gridTemplateAreas", "myArea myArea . . ."],
] as const;

const backgroundProps = [
  [
    "background",
    "background",
    "lightblue url('foo.jpg') no-repeat fixed center",
  ],
  ["backgroundImage", "background-image", "url(foo.jpg)"],
  ["backgroundSize", "background-size", "center"],
  ["backgroundRepeat", "background-repeat", "no-repeat"],
] as const;

const positionProps = [
  ["top", "0px"],
  ["bottom", "0px"],
  ["right", "0px"],
  ["left", "0px"],
  ["position", "fixed"],
  ["position", "sticky"],
  ["position", "absolute"],
  ["position", "static"],
  ["position", "relative"],
] as const;

export const getDefaultValue = (value?: string | number) => {
  const spaceArrayLength = space.length - 1;
  if (value === undefined) return value;
  const parsedValue = +value;
  if (typeof value === "string" && parsedValue > spaceArrayLength) {
    return `${value}px`;
  }
  if (parsedValue <= spaceArrayLength) {
    return space[parsedValue];
  }
  if (parsedValue > spaceArrayLength) {
    return `${parsedValue * 8}px`;
  }
  return value;
};

const testStyledSystemGrid = (
  component: (gridProperties?: GridProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(gridProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component(props));

        assertStyleMatch({ [propName]: value }, elementQuery());
      });
    },
  );
};

const testStyledSystemBackground = (
  component: (backgroundProperties?: BackgroundProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(backgroundProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));

        assertStyleMatch({ [styledSystemProp]: value }, elementQuery());
      });
    },
  );
};

const testStyledSystemPosition = (
  component: (positionProperties?: PositionProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(positionProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, value) => {
      it(`then ${styledSystemProp} should have been set correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));

        assertStyleMatch({ [styledSystemProp]: value }, elementQuery());
      });
    },
  );
};

// this util will catch that a console output occurred without polluting the output when running the unit tests
const expectConsoleOutput = (
  message: string,
  type: "warn" | "error" = "error",
) => {
  if (!message) {
    throw new Error(`no ${type} message provided`);
  }

  expect.assertions(1);

  const consoleType = global.console[type];
  let consoleArgs: string[];

  jest.spyOn(global.console, type).mockImplementation((...args) => {
    if (!args.length) return;

    const msg = args.join(" ");
    const params = args.slice(1, args.length);

    if (sprintf(msg, ...params).includes(message)) {
      consoleArgs = args;
      return;
    }

    consoleType(...args);
  });

  return () => {
    if (consoleArgs) {
      // eslint-disable-next-line no-console
      expect(console[type]).toHaveBeenCalledWith(...consoleArgs);
    }

    global.console[type] = consoleType;
  };
};

const testStyledSystemMargin = (
  component: (spacingProps?: MarginProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: MarginProps,
  assertOpts?: jest.Options,
) => {
  describe("default props", () => {
    let StyleElement: HTMLElement;

    beforeAll(() => {
      render(component({ ...defaults }));
      StyleElement = elementQuery();
    });

    it("should set the correct margins", () => {
      let margin;
      let marginLeft;
      let marginRight;
      let marginTop;
      let marginBottom;

      if (defaults) {
        margin = getDefaultValue(defaults.m);
        marginLeft = getDefaultValue(defaults.ml || defaults.mx);
        marginRight = getDefaultValue(defaults.mr || defaults.mx);
        marginTop = getDefaultValue(defaults.mt || defaults.my);
        marginBottom = getDefaultValue(defaults.mb || defaults.my);

        assertStyleMatch(
          {
            margin,
            marginLeft,
            marginRight,
            marginTop,
            marginBottom,
          },
          StyleElement,
          assertOpts,
        );
      } else {
        expect(StyleElement).not.toHaveStyleRule("marginLeft");
        expect(StyleElement).not.toHaveStyleRule("marginRight");
        expect(StyleElement).not.toHaveStyleRule("marginTop");
        expect(StyleElement).not.toHaveStyleRule("marginBottom");
        expect(StyleElement).not.toHaveStyleRule("margin");
      }
    });
  });

  describe.each(marginProps)(
    'when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: 2 };
        render(component({ ...props }));

        assertStyleMatch(
          { [propName]: "var(--spacing200)" },
          elementQuery(),
          assertOpts,
        );
      });
    },
  );
};

const testStyledSystemPadding = (
  component: (spacingProps?: PaddingProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: PaddingProps,
  assertOpts?: jest.Options,
) => {
  describe("default props", () => {
    let StyleElement: HTMLElement;

    beforeAll(() => {
      render(component({ ...defaults }));
      StyleElement = elementQuery();
    });

    it("should set the correct paddings", () => {
      let padding;
      let paddingLeft;
      let paddingRight;
      let paddingTop;
      let paddingBottom;

      if (defaults) {
        padding = getDefaultValue(defaults.p);
        paddingLeft = getDefaultValue(defaults.pl || defaults.px);
        paddingRight = getDefaultValue(defaults.pr || defaults.px);
        paddingTop = getDefaultValue(defaults.pt || defaults.py);
        paddingBottom = getDefaultValue(defaults.pb || defaults.py);

        assertStyleMatch(
          {
            padding,
            paddingLeft,
            paddingRight,
            paddingTop,
            paddingBottom,
          },
          StyleElement,
          assertOpts,
        );
      } else {
        expect(StyleElement).not.toHaveStyleRule("paddingLeft");
        expect(StyleElement).not.toHaveStyleRule("paddingRight");
        expect(StyleElement).not.toHaveStyleRule("paddingTop");
        expect(StyleElement).not.toHaveStyleRule("paddingBottom");
        expect(StyleElement).not.toHaveStyleRule("padding");
      }
    });
  });

  describe.each(paddingProps)(
    'when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: 2 };
        render(component({ ...props }));

        assertStyleMatch(
          { [propName]: "var(--spacing200)" },
          elementQuery(),
          assertOpts,
        );
      });
    },
  );
};

const testStyledSystemSpacing = (
  component: (spacingProps?: MarginProps | PaddingProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: MarginProps | PaddingProps,
  assertOpts?: jest.Options,
) => {
  testStyledSystemMargin(
    component,
    elementQuery,
    defaults as MarginProps,
    assertOpts,
  );
  testStyledSystemPadding(
    component,
    elementQuery,
    defaults as PaddingProps,
    assertOpts,
  );
};

const testStyledSystemColor = (
  // https://stackoverflow.com/questions/53711454/styled-system-props-typing-with-typescript
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  component: (colorProperties?: any) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(colorProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));
        const StyleElement = elementQuery();
        // Some props need to have camelcase so used toHaveStyleRule rather than assertStyleMatch
        expect(StyleElement).toHaveStyleRule(propName, value);
      });
    },
  );
};

const testStyledSystemWidth = (
  component: (widthProperties?: { width: string }) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe("when a width prop is specified using styled system props", () => {
    it("should set the width styling correctly", () => {
      const [styledSystemProp, propName, value] = widthProps;
      const props = { [styledSystemProp]: value };
      render(component({ ...props }));
      const StyleElement = elementQuery();
      expect(StyleElement).toHaveStyleRule(propName, value);
    });
  });
};

const testStyledSystemHeight = (
  component: (heightProperties?: { height: string }) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe("when a height prop is specified using styled system props", () => {
    it("should set the height styling correctly", () => {
      const [styledSystemProp, propName, value] = heightProps;
      const props = { [styledSystemProp]: value };
      render(component({ ...props }));
      const StyleElement = elementQuery();
      expect(StyleElement).toHaveStyleRule(propName, value);
    });
  });
};

const testStyledSystemLayout = (
  component: (layoutProperties?: LayoutProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(layoutProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));
        const StyleElement = elementQuery();
        // Some props need to have camelcase so used toHaveStyleRule rather than assertStyleMatch
        expect(StyleElement).toHaveStyleRule(propName, value);
      });
    },
  );
};

const testStyledSystemFlexBox = (
  component: (flexboxProperties?: FlexboxProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(flexBoxProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component(props));

        assertStyleMatch({ [propName]: value }, elementQuery());
      });
    },
  );
};

export {
  assertStyleMatch,
  toCSSCase,
  makeArrayKeys,
  mockMatchMedia,
  testStyledSystemColor,
  testStyledSystemLayout,
  testStyledSystemFlexBox,
  testStyledSystemGrid,
  testStyledSystemBackground,
  testStyledSystemPosition,
  expectConsoleOutput,
  testStyledSystemSpacing,
  testStyledSystemMargin,
  testStyledSystemPadding,
  testStyledSystemWidth,
  testStyledSystemHeight,
};
