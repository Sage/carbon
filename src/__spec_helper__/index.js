import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { setup } from './mock-match-media';

setup();
Enzyme.configure({ adapter: new Adapter() });
