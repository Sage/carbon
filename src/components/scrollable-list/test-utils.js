import React from 'react';
import { makeArrayKeys } from '../../__spec_helper__/test-utils';
import ScrollableListItem from './scrollable-list-item.component';

const listItemReducer = ({ nonSelectables = [], customSelectables = [] }) => {
  // generate jsx for selectable list items, based on indexes in config
  return (acc, item, index) => {
    if (nonSelectables.includes(item)) return [...acc, <div />];
    if (customSelectables.includes(item)) return [...acc, <div isSelectable />];

    return [...acc, <ScrollableListItem id={ index } />];
  };
};

const renderListItems = (opts) => {
  return makeArrayKeys(opts.num).reduce(listItemReducer(opts), []);
};

export default renderListItems;
