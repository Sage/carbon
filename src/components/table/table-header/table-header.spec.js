import React from 'react';
import 'jest-styled-components';
import TestUtils from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import { Table, TableRow } from '..';
import TableHeader from '.';
import StyledTableHeader from './table-header.style';
import { rootTagTest } from '../../../utils/helpers/tags/tags-specs';
import BaseTheme from '../../../style/themes/base';
import ClassicTheme from '../../../style/themes/classic';
import { assertStyleMatch } from '../../../__spec_helper__/test-utils';
import tableSizes from '../table-sizes.style';

describe('TableHeader', () => {
  let instance, instanceSortable, sortableColumn, sortableHeader, changeSpy;

  beforeEach(() => {
    changeSpy = jasmine.createSpy('changeSpy');

    instance = TestUtils.renderIntoDocument(
      <Table>
        <TableRow>
          <TableHeader
            align='right'
            style={ { width: '50px' } }
          />
        </TableRow>
      </Table>
    );

    instanceSortable = TestUtils.renderIntoDocument(
      <Table onChange={ changeSpy }>
        <TableRow>
          <TableHeader sortable name='name' />
        </TableRow>
      </Table>
    );

    sortableColumn = TestUtils.findRenderedDOMComponentWithTag(instanceSortable, 'th');
    sortableHeader = TestUtils.scryRenderedComponentsWithType(instanceSortable, TableHeader)[0];
  });

  describe('prop checking', () => {
    beforeEach(() => {
      jest.spyOn(global.console, 'error').mockImplementation(() => {});
    });

    it('throws an error if no name prop is passed', () => {
      const expected = 'Sortable columns require a prop of name of type String';

      mount(
        <Table onChange={ changeSpy } sortOrder='desc'>
          <TableRow>
            <TableHeader sortable />
          </TableRow>
        </Table>
      );


      expect(console.error.mock.calls[0][0]).toMatch(expected);
    });

    it('throws an error if the name is not a string', () => {
      const expected = 'Failed prop type: name must be a string';

      mount(
        <Table onChange={ changeSpy } sortOrder='desc'>
          <TableRow>
            <TableHeader sortable name={ 123 } />
          </TableRow>
        </Table>
      );

      expect(console.error.mock.calls[0][0]).toMatch(expected);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });
  });

  describe('emitSortEvent', () => {
    describe('if no sortOrder has been specified', () => {
      it('calls the tables onSort function with the default params', () => {
        TestUtils.Simulate.click(sortableColumn);
        expect(instanceSortable.props.onChange).toHaveBeenCalledWith(
          'table', {
            currentPage: '',
            pageSize: '',
            filter: {},
            sortedColumn: 'name',
            sortOrder: 'desc'
          }
        );
      });

      describe('if the column has already has been sorted', () => {
        it('flips the sortOrder when clicked', () => {
          sortableHeader.context.sortedColumn = 'name';
          sortableHeader.context.sortOrder = 'asc';
          TestUtils.Simulate.click(sortableColumn);
          expect(instanceSortable.props.onChange).toHaveBeenCalledWith(
            'table', {
              currentPage: '',
              filter: {},
              pageSize: '',
              sortedColumn: 'name',
              sortOrder: 'desc'
            }
          );
        });
      });
    });

    describe('if a sortOrder has been passed', () => {
      it('sends the passed in sortOrder prop to the onSort function', () => {
        sortableHeader.context.sortOrder = 'desc';
        TestUtils.Simulate.click(sortableColumn);
        expect(instanceSortable.props.onChange).toHaveBeenCalledWith(
          'table', {
            currentPage: '',
            filter: {},
            pageSize: '',
            sortedColumn: 'name',
            sortOrder: 'desc'
          }
        );
      });
    });
  });

  describe('sortable columns', () => {
    describe('if a column is sortable', () => {
      let ariaLink;

      beforeEach(() => {
        ariaLink = sortableColumn.querySelector('a[aria-label]');
      });

      describe('aria-label', () => {
        let wrapper;

        beforeEach(() => {
          wrapper = mount(
            <Table onChange={ changeSpy }>
              <TableRow>
                <TableHeader sortable name='name' />
              </TableRow>
            </Table>
          );
        });

        it('states the column is sortable, and activating will sort descending', () => {
          ariaLink = wrapper.find('a[aria-label="Sortable column, activate to sort column descending"]');
          expect(ariaLink.exists()).toBe(true);
        });

        it('includes the current sort order after sorting', () => {
          wrapper.setProps({ sortOrder: 'desc' });
          ariaLink = wrapper
            .find('a[aria-label="Sortable column, sorted descending, activate to sort column ascending"]');
          expect(ariaLink.exists()).toBe(true);
        });
      });

      describe('aria-sort', () => {
        let wrapper;

        beforeEach(() => {
          wrapper = mount(
            <Table onChange={ changeSpy }>
              <TableRow>
                <TableHeader sortable name='name' />
                <TableHeader sortable name='age' />
              </TableRow>
            </Table>
          );
        });

        it('does not exist until after a column has been sorted', () => {
          const header = wrapper.find('th[aria-sort]');
          expect(header.exists()).toBe(false);
        });

        it('is set to ascending when the name column is sorted ascending', () => {
          wrapper.setProps({
            sortedColumn: 'name',
            sortOrder: 'asc'
          });

          const header = wrapper.find('th[aria-sort]');
          expect(header.length).toEqual(1);
          expect(header.props()['aria-sort']).toEqual('ascending');
        });

        it('is set to descending when the age column is sorted descending', () => {
          wrapper.setProps({
            sortedColumn: 'age',
            sortOrder: 'desc'
          });

          const header = wrapper.find('th[aria-sort="descending"]');
          expect(header.length).toEqual(1);
          expect(header.props()['aria-sort']).toEqual('descending');
        });
      });

      describe('before a sortable header is clicked', () => {
        it('does not display an icon', () => {
          expect(sortableHeader.sortIconHTML).toEqual(null);
        });

        it('contains a link with a descriptive aria-label that defaults to descending', () => {
          expect(ariaLink.getAttribute('aria-label')).toEqual('Sortable column, activate to sort column descending');
        });
      });

      describe('after a sortable header has been clicked', () => {
        describe('when the sortOrder is descending', () => {
          beforeEach(() => {
            sortableHeader.context.sortedColumn = 'name';
            sortableHeader.context.sortOrder = 'desc';
            TestUtils.Simulate.click(sortableColumn);
          });

          it('adds the sort_down icon', () => {
            expect(sortableHeader.sortIconHTML.props.type).toEqual('sort_down');
          });
        });

        describe('when the sortOrder is ascending or not specified', () => {
          it('adds the sort_up icon', () => {
            sortableHeader.context.sortedColumn = 'name';
            TestUtils.Simulate.click(sortableColumn);
            expect(sortableHeader.sortIconHTML.props.type).toEqual('sort_up');
          });
        });
      });
    });

    describe('if a column is not sortable', () => {
      it('does not return an icon', () => {
        const nonSortableHeader = TestUtils.scryRenderedComponentsWithType(instance, TableHeader)[0];
        expect(nonSortableHeader.sortIconHTML).toEqual(null);
      });

      it('does not contain an aria-label element', () => {
        const wrapper = mount(
          <Table onChange={ changeSpy }>
            <TableRow>
              <TableHeader sortable={ false } name='name' />
            </TableRow>
          </Table>
        );

        const ariaLabel = wrapper.find('a[aria-label]');
        expect(ariaLabel.length).toEqual(0);
      });
    });

    describe('sortDescription', () => {
      describe('when a column is sortable', () => {
        let wrapper;
        let context;

        beforeEach(() => {
          context = {
            sortOrder: null
          };
          wrapper = shallow(<TableHeader sortable name='name' />, { context });
        });

        it('says the column is a "Sortable column, activate to sort column descending"', () => {
          expect(wrapper.instance().sortDescription).toEqual('Sortable column, activate to sort column descending');
        });

        it('includes "sorted descending"  when the current sort order is "desc"', () => {
          wrapper.setContext({ sortOrder: 'desc' });
          expect(
            wrapper.instance().sortDescription
          ).toEqual('Sortable column, sorted descending, activate to sort column ascending');
        });

        it('includes "sorted ascending"  when the current sort order is "asc"', () => {
          wrapper.setContext({ sortOrder: 'asc' });
          expect(
            wrapper.instance().sortDescription
          ).toEqual('Sortable column, sorted ascending, activate to sort column descending');
        });
      });

      describe('when a column is not sortable', () => {
        it('returns null', () => {
          const wrapper = shallow(<TableHeader sortable={ false } name='name' />);
          expect(wrapper.instance().sortDescription).toBeNull();
        });
      });
    });

    describe('onSortableColumnClick', () => {
      it('calls event.preventDefault', () => {
        const wrapper = shallow(<TableHeader sortable name='name' />);
        const link = wrapper.find('a[href]');

        const mockEvent = {
          preventDefault() {
            console.log('mockEvent.preventDefault() called');
          }
        };

        spyOn(mockEvent, 'preventDefault');

        link.simulate('click', mockEvent);
        expect(mockEvent.preventDefault).toHaveBeenCalled();
      });
    });
  });

  describe.each(['compact', 'small', 'medium', 'large'])(
    'when the theme is classic',
    (size) => {
      const wrapper = mount(
        <Table>
          <TableRow>
            <TableHeader
              theme={ ClassicTheme }
              size={ size }
            />
          </TableRow>
        </Table>
      );

      it(`matches the expected style when the size is ${size}`, () => {
        assertStyleMatch({
          height: tableSizes.medium.height
        }, wrapper.find(StyledTableHeader));
      });
    },
  );

  describe('render', () => {
    let wrapper, th;
    beforeEach(() => {
      wrapper = mount(
        <Table>
          <TableRow>
            <TableHeader
              align='right'
              style={ { width: '50px' } }
            />
          </TableRow>
        </Table>
      );
      th = wrapper.find('th').hostNodes();
    });

    it('renders additional props to the th element', () => {
      expect(th.exists()).toBeTruthy();
      expect(th.prop('style').width).toEqual('50px');
    });

    it('renders a th to match the expected default style', () => {
      assertStyleMatch({
        backgroundColor: BaseTheme.table.header,
        borderBottom: `1px solid ${BaseTheme.table.secondary}`,
        borderLeft: `1px solid ${BaseTheme.colors.border}`,
        color: BaseTheme.colors.white,
        textAlign: 'right'
      }, th);
    });

    it('renders the sort icon to match the snapshot', () => {
      wrapper = mount(
        <Table>
          <TableRow>
            <TableHeader
              align='right'
              sortable
            />
          </TableRow>
        </Table>
      );
      th = wrapper.find('th').hostNodes();
      expect(th).toMatchSnapshot();
    });

    it('renders a th to match the expected default style when a width is provided', () => {
      wrapper = mount(
        <Table>
          <TableRow>
            <TableHeader
              align='right'
              sortable
              width='200'
            />
          </TableRow>
        </Table>
      );
      th = wrapper.find('th').hostNodes();

      assertStyleMatch({
        width: '200px'
      }, th);
    });
  });

  describe('tags on component', () => {
    const wrapper = shallow(<TableHeader data-element='bar' data-role='baz' />);

    it('include correct component, element and role data tags', () => {
      rootTagTest(wrapper, 'table-header', 'bar', 'baz');
    });
  });
});
