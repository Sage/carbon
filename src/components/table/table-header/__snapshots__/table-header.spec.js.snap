// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHeader render renders the sort icon to match the snapshot 1`] = `
.c1 {
  background-color: #335B6D;
  border-width: 0;
  border-bottom: 1px solid #CCD6DA;
  border-left: 1px solid #668491;
  box-sizing: border-box;
  color: #FFFFFF;
  font-weight: 700;
  outline: medium none;
  padding: 0 8px;
  position: relative;
  text-align: right;
  text-overflow: ellipsis;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  white-space: nowrap;
  cursor: pointer;
}

.c1:first-child {
  border-left: none;
}

.c1:hover {
  background-color: #19475B;
}

.c1 .c3 {
  color: #FFFFFF;
}

.c1 a:link,
.c1 a:visited,
.c1 a:hover,
.c1 a:active {
  color: #FFFFFF;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  line-height: 16px;
  height: 24px;
}

.c1 a span {
  float: left;
}

.c1 a span .carbon-icon__svg--sort-down,
.c1 a span .carbon-icon__svg--sort-up {
  height: 11px;
  width: 10px;
}

.c2 .c5 .c4,
.c2 .c5 .c0 {
  font-size: 14px;
  padding-left: 11px;
  padding-right: 11px;
}

<th
  className="c0 c1"
  data-component="table-header"
  onClick={[Function]}
>
  <a
    aria-label="Sortable column, activate to sort column descending"
    href="#sort-asc"
    onClick={[Function]}
  />
</th>
`;
